File: core/src/main/java/no/elg/infiniteBootleg/util/Metadata.java
package no.elg.infiniteBootleg.util;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * @author Elg
 */
public interface Metadata {

  @Nullable
  default <T> T get(@NotNull String key) {
    return get(key, null);
  }

  /**
   * @param key Key to the castle
   * @param defaultVal Value to return if no object with the given key exists
   * @param <T> The type of the value
   * @return The value stored at {@code key}
   * @throws IllegalStateException If the object at {@code key} cannot be cast to {@code T}
   */
  @Nullable
  <T> T get(@NotNull String key, @Nullable T defaultVal);

  <T> void set(@NotNull String key, T value);
}

File: core/src/main/kotlin/no/elg/infiniteBootleg/api/Metadata.kt
package no.elg.infiniteBootleg.api

/**
 * @author Elg
 */
interface Metadata {
  operator fun <T> get(key: String): T? {
    return get<T?>(key, null)
  }

  /**
   * @param key Key to the castle
   * @param defaultVal Value to return if no object with the given key exists
   * @param <T> The type of the value
   * @return The value stored at `key`
   * @throws IllegalStateException If the object at `key` cannot be cast to `T`
   </T> */
  operator fun <T> get(key: String, defaultVal: T?): T?
  operator fun <T> set(key: String, value: T)
}
